### PR TITLE
Added documentation to rcl_lifecycle

### PR DESCRIPTION
**Pull request by ahcorde**
Added documentation to:

 - rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h
 - rcl_lifecycle/include/rcl_lifecycle/default_state_machine.h
 - rcl_lifecycle/include/rcl_lifecycle/transition_map.h

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

File: rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h [hunk 1]
```diff
@@ -26,10 +44,41 @@ extern "C"
...
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[inout] state pointer to the state struct to be initialized
+ * \param[in] id identifier of the state
+ * \param[in] label label of the state
+ * \param[in] allocator a valid allocator used to initialized the lifecycle state
+ * \return `RCL_RET_OK` if state was initialized successfully, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h:80]: Through the whole code of `rcl_lifecycle.c` invalid (NULL) parameters are handled as `RCL_RET_ERROR` and these could be returned as `RCL_INVALID_ARGUMENT`, I think we could add a PR to handle this.

File: rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h [hunk 2]
```diff
@@ -39,23 +88,70 @@ rcl_lifecycle_state_init(
   const char * label,
   const rcl_allocator_t * allocator);
 
+/// Finalize a rcl_lifecycle_state_t.
+/**
+ *
+ * Calling this will set the rcl_lifecycle_state_t struct into the an unitialized state that is
+ * functionally the same as before rcl_lifecycle_state_init was called.
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h:95]: This is truth but I think the wording should be changed to explicitly note that this method will "close" the structure by freeing the resources taken by the init allocator.

Comment (Blast545) [reply rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h:95]: Something similar to this [rcl timer example](https://github.com/ros2/rcl/blob/8a0ec5fe959680c18e599fb4580813ddf14fe352/rcl/include/rcl/timer.h#L163)

File: rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h [hunk 3]
```diff
@@ -39,23 +88,70 @@ rcl_lifecycle_state_init(
...
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[inout] state struct to be deinitialized
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h:105]:
```suggestion
 * \param[inout] state struct to be finalized
```

File: rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h [hunk 4]
```diff
@@ -39,23 +88,70 @@ rcl_lifecycle_state_init(
...
+ * \param[in] allocator a valid allocator used to deinitialized the lifecycle state
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h:106]:
```suggestion
 * \param[in] allocator a valid allocator used to finalize the lifecycle state
```

File: rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h [hunk 5]
```diff
@@ -39,23 +88,70 @@ rcl_lifecycle_state_init(
...
+ * \return `RCL_RET_OK` if the state was deinitialized successfully, or
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h:107]:
```suggestion
 * \return `RCL_RET_OK` if the state was finalized successfully, or
```

File: rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h [hunk 6]
```diff
@@ -68,17 +164,71 @@ rcl_lifecycle_transition_init(
   rcl_lifecycle_state_t * goal,
   const rcl_allocator_t * allocator);
 
+/// Finalize a rcl_lifecycle_transition_t.
+/**
+ * Calling this will set the rcl_lifecycle_transition_t struct into the an unitialized state that is
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h:169]: Same suggestions as in L95

File: rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h [hunk 7]
```diff
@@ -68,17 +164,71 @@ rcl_lifecycle_transition_init(
...
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[inout] transition struct to be deinitialized
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h:180]: Suggestion: changing deinitialized to finalized when using `_fini` methods

File: rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h [hunk 8]
```diff
@@ -102,26 +271,99 @@ rcl_lifecycle_state_machine_fini(
...
 RCL_LIFECYCLE_PUBLIC
 RCL_WARN_UNUSED
 const rcl_lifecycle_transition_t *
 rcl_lifecycle_get_transition_by_label(
   const rcl_lifecycle_state_t * state,
   const char * label);
 
+/// Trigger a state by id.
+/**
+ * This function will trigger a transition base on the `id`. If the argument
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h:348]:
```suggestion
 * This function will trigger a transition based on the `id`. If the argument
```

File: rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h [hunk 9]
```diff
@@ -102,26 +271,99 @@ rcl_lifecycle_state_machine_fini(
...
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] state_machine pointer to the state machine struct
+ * \param[in] id identifier of the transition to be triggered
+ * \param[in] publish_notification if the value is `true` a message will be published
+ *    notifying the transition, otherwise no message will be published
+ * \return `RCL_RET_OK` if the transition was trigger successfully, or
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h:364]:
```suggestion
 * \return `RCL_RET_OK` if the transition was triggered successfully, or
```

File: rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h [hunk 10]
```diff
@@ -130,6 +372,27 @@ rcl_lifecycle_trigger_transition_by_id(
...
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] state_machine pointer to the state machine struct
+ * \param[in] label of the transition to be triggered
+ * \param[in] publish_notification if the value is `true` a message will be published
+ *    notifying the transition, otherwise no message will be published
+ * \return `RCL_RET_OK` if the transition was trigger successfully, or
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h:393]:
```suggestion
 * \return `RCL_RET_OK` if the transition was triggered successfully, or
```

File: rcl_lifecycle/include/rcl_lifecycle/transition_map.h [hunk 11]
```diff
@@ -25,24 +25,87 @@ extern "C"
...
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
 RCL_LIFECYCLE_PUBLIC
 RCL_WARN_UNUSED
 rcl_ret_t
 rcl_lifecycle_transition_map_fini(
   rcl_lifecycle_transition_map_t * transition_map,
   const rcl_allocator_t * allocator);
 
+/// Regiter a state
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/transition_map.h:90]:
```suggestion
/// Register a state
```

File: rcl_lifecycle/include/rcl_lifecycle/transition_map.h [hunk 12]
```diff
@@ -25,24 +25,87 @@ extern "C"
...
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] map
+ * \param[in] state the state to register
+ * \param[in] allocator a valid allocator used to deinitialized the state machine
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/transition_map.h:104]: I think it reallocates it, I wouldn't use deinitialize in this context, maybe "used to register" or "used to reallocate"

File: rcl_lifecycle/include/rcl_lifecycle/transition_map.h [hunk 13]
```diff
@@ -51,6 +114,25 @@ rcl_lifecycle_register_state(
   rcl_lifecycle_state_t state,
   const rcl_allocator_t * allocator);
 
+/// Regiter a transition
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/transition_map.h:117]:
```suggestion
/// Register a transition
```

File: rcl_lifecycle/include/rcl_lifecycle/transition_map.h [hunk 14]
```diff
@@ -51,6 +114,25 @@ rcl_lifecycle_register_state(
...
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] transition_map
+ * \param[in] transition the transition to register
+ * \param[in] allocator a valid allocator used to deinitialized the state machine
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/transition_map.h:131]: Same as above

File: rcl_lifecycle/include/rcl_lifecycle/transition_map.h [hunk 15]
```diff
@@ -51,6 +114,25 @@ rcl_lifecycle_register_state(
...
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] transition_map
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/transition_map.h:129]: Add brief description saying along the lines of "transition_map to be modified". Here and below

File: rcl_lifecycle/include/rcl_lifecycle/transition_map.h [hunk 16]
```diff
@@ -59,13 +141,45 @@ rcl_lifecycle_register_transition(
...
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] transition_map
+ * \param[in] state_id
```
Comment (Blast545) [inline rcl_lifecycle/include/rcl_lifecycle/transition_map.h:180]:
```suggestion
 * \param[in] state_id used to get the label to search
```

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10484/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5961/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8545/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10362/)

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10514/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5981/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8566/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10388/)

Unrelated warning

**Pull request merged**